docs/designs/allreduce.md

```diff
@@ -1,14 +1,198 @@
+that we mentioned in the previous section. The main differences are the following:
+
+1. Gradients from each worker are not sent to master.
+1. The average of gradients across all workers is calculated locally on each worker.
+1. The model is updated directly on each worker and each worker has the exact same copy of the model.
+
+Below is the pseudo-code for this process on each worker:
+
+```python
+communicator = AllreduceCommunicator()
+
+with tf.GradientTape() as tape:
+    outputs = model(features, training=True)
+    loss = loss_fn(outputs, labels)
+local_grads = tape.gradient(loss, get_trainable_items())
+res, averaged_grads = communicator.average_gradients(grads)
+if res == SUCCESS:
+    update_model(averaged_grads)
+else:
+    report_failure()
+    continue
+```
+
+#### Failure Handling during Training
+
+The above pseudo-code will be wrapped and executed for each batch of the dataset inside `process_minibatch_and_report()`
+in the code below. Each worker continues to perform tasks until there is no new batch available.
+
+```python
+while True:
+    dataset = self._task_data_service.get_dataset()
+    if not dataset:
+        break
+    dataset = dataset.batch(self._minibatch_size).prefetch(1)
+    for dataset_batch in dataset:
+        task = self._task_data_service.get_current_task()
+        process_minibatch_and_report(dataset_batch, task)
+```
+
+If any of the gradients averaging operation fails while the workers are still active and healthy, we simply report the
+failure and continue training on the next batch.
+
+If any of the workers fails, e.g. the pod is accidentally killed, the following steps will be performed:
+
+1. The task that the failed worker was handling will be added back to the task queue.
+1. Since the `AllreduceCommunicator` is aware of the failure. It will reconstruct the communicator and re-assign ranks among
+the existing active workers. The existing workers will then continue to run on the tasks at hand.
+1. The master pod will try to create a new worker pod after a specified restart delay period.
+1. Once the worker pod becomes active and `AllreduceCommunicator` is aware of it, we perform the following steps:
+    1. Lock all existing worker pods.
+    1. Send the current model from one of the worker pods to master pod.
+    1. Initialize the model on the new worker pod using the current model that master pod just received.
+    1. All workers continue to operate and perform allreduce operations to update the model.
+
+Note that since the existing active workers have the exact same copy of the model after the previous allreduce operation completes,
+we can guarantee that the new worker will have the same copy of the model as the ones on other workers once the next
+allreduce operation completes.
+
+#### Training with Evaluation
+
+If the worker encounters any evaluation tasks in the above process, it will evaluate the model directly on master once the
```

Comment:
Need to revise this sentence... `it will evaluate the model directly on master` --> looks like the master will evaluate the model, but actually workers use the model received from the master..

Reply:
@ywskycn This is a typo. I changed it to "it will evaluate the model directly on the **worker**". 

@QiJune Since we currently have master to perform TensorBoard and start new workers, etc. We still need to send the evaluation results to master for TensorBoard visualization purposes. ElasticDL needs to support master election in the future so we can reuse one of the worker pods as master pod. I'll open a separate issue for that.